### PR TITLE
Make use of Gradle Module Metadata

### PR DESCRIPTION
**Pull request by Su5eD**
This PR simplifies the KFF setup process using Gradle [module metadata](https://docs.gradle.org/current/userguide/publishing_gradle_module_metadata.html) and [variant aware selection](https://docs.gradle.org/current/userguide/variant_model.html).

## The issue

Currently, KFF's setup is a little overcomplicated. The setup guide uses the legacy plugin application api, and also instructs you to apply a remote script in your build.gradle. Kotlin is added to the `minecraft_classpath` and KFF's maven dependencies are ignored. This might appear confusing, especially to beginners.

## The solution

Introducing: Gradle Module Metadata, making dependency management simpler. With GMM, we can instruct gradle to use separate sets of artifacts and maven dependencies when added into a project. In our case, we want to put kotlin and the subproject jars onto the compile classpath, then use a fat jar with all dependencies included at runtime.

#### KFFMod subproject

The root project KFF mod has been moved into its own subproject so that we can put the jarJar onto the `PLUGIN` layer.

### Layer layout

We bundle the 3 subproject's artifacts - `kfflang`, `kfflib` and `kffmod` into one fat jar using `jarJar`.  
Kotlin dependencies remain shaded due to jarjar's default layer selection. For more information, see:

- https://github.com/MinecraftForge/MinecraftForge/issues/8878

The fat jar needs to be added to the `PLUGIN` layer using an `FMLModType: LIBRARY` manifest attribute, so that our embedded jars can access kotlin classes.  
The `kfflang` jar contains a language provider and is therefore placed on the `PLUGIN` layer, too.  
The remaining 2 jars are put on the `GAME` layer.

To avoid a mod conflict between the main jar and kffmod (both jars are named `kotlinforforge`), each non-mod jar is given an `Automatic-Module-Name`:
- `<root>` -> `thedarkcolour.kotlinforforge`
- `kfflang` -> `thedarkcolour.kotlinforforge.lang`
- `kfflib` -> `thedarkcolour.kotlinforforge.lib`

### Running the main project

ForgeGradle sets the minecraft runs' `MOD_CLASSES` environment variable after the project has been evaluated, which makes it a bit tricky to remove it. And even if you manage to do that, forge will default to an empty string and try to turn that into a real Path - which doesn't exists - and crashed the game.

The easiest workaround to this is simply forcing the creation of the project output directories, so that `MOD_CLASSES` is not empty but also contains existing paths.
```gradle
whenTaskAdded {  
    if (name == "prepareRuns") {  
        doFirst {  
            sourceSets.main.get().output.files.forEach(File::mkdirs)  
        }  
    }  
}
```

## The new, improved setup

With all these changes in place, the setup is made much easier.
```gradle
plugins {
    // Adds the Kotlin Gradle plugin
    id 'org.jetbrains.kotlin.jvm' version '1.7.20'
    // OPTIONAL Kotlin Serialization plugin
    id 'org.jetbrains.kotlin.plugin.serialization' version '1.7.20'
}

repositories {
    // Add KFF Maven repository
    maven {
        name = 'Kotlin for Forge'
        url = 'https://thedarkcolour.github.io/KotlinForForge/'
    }
}

dependencies {
    // Adds KFF as dependency and Kotlin libs
    implementation 'thedarkcolour:kotlinforforge:3.8.0'
}
```

We can see the dependencies were applied to the correct scopes:
dependency scopes

## Dependency Diagram

```mermaid
graph LR
    MAIN[thedarkcolour.kotlinforforge]

    subgraph Subprojects
        LANG[thedarkcolour.kotlinforforge.lang]
        LIB[thedarkcolour.kotlinforforge.lib]
        MOD[kotlinforforge]
    end

    MAIN --> LANG
    MAIN --> LIB
    MAIN --> MOD

    MOD --> LANG
    MOD --> LIB

    LANG --> STDLIB
    LANG --> STDLIB_COMMON
    LANG --> STDLIB_JDK8
    LANG --> STDLIB_JDK7

    LIB --> LANG
    LIB --> STDLIB_JDK8
    LIB --> REFLECT
    LIB --> COROUTINES_CORE
    LIB --> COROUTINES_CORE_JVM
    LIB --> COROUTINES_JDK8
    LIB --> SERIALIZATION_JSON

    MOD --> STDLIB
    MOD --> STDLIB_COMMON
    MOD --> STDLIB_JDK8
    MOD --> STDLIB_JDK7
    MOD --> REFLECT

    subgraph Kotlin
        STDLIB[kotlin-stdlib]
        STDLIB_COMMON[kotlin-stdlib-common]
        STDLIB_JDK8[kotlin-stdlib-jdk8]
        STDLIB_JDK7[kotlin-stdlib-jdk7]
        REFLECT[kotlin-relfect]
        COROUTINES_CORE[kotlinx-coroutines-core]
        COROUTINES_CORE_JVM[kotlinx-coroutines-core-jvm]
        COROUTINES_JDK8[kotlinx-coroutines-jdk8]
        SERIALIZATION_JSON[kotlinx-serialization-json]
    end
```

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -1,72 +1,151 @@
...
 
-// Adds to JarJar without using as Gradle dependency
-fun DependencyHandlerScope.compileLibrary(group: String, name: String, version: String, maxVersion: String, obf: Boolean = false) {
-    val lib = this.create(group, name, version = "[$version,$maxVersion)")
-    jarJar(lib) {
-        isTransitive = false
-        jarJar.pin(this, version)
+tasks {
+    jar {
+        isEnabled = false
```
Comment (thedarkcolour) [inline build.gradle.kts]: This causes a build error when kotlinforforge-x.x.x.jar doesn't exist in the build/libs folder

Comment (Su5eD) [reply build.gradle.kts]: fixed

**Comment by tmvkrpxl0** [comment]
image
This doesn't feel right

**Comment by tmvkrpxl0** [comment]
@Su5eD Could this be solved easily? It really wouldn't cause any problem, but It just feels weird to see those. If it couldn't then we will ship with those.

**Comment by tmvkrpxl0** [comment]
> This doesn't feel right

That is from kotlin-reflect so we really can't remove it

**Comment by thedarkcolour** [comment]
Everything works with my testing 👍

**Pull request merged**